apps/emqx_management/src/emqx_mgmt_api_clients.erl

```diff
@@ -851,6 +838,145 @@ do_unsubscribe(ClientID, Topic) ->
+                {enough, NResultAcc} ->
+                    %% TODO: add persistent session count?
+                    FQueryState = emqx_mgmt_api:maybe_collect_total_from_tail_nodes(
+                        Tail, NQueryState
+                    ),
+                    FComplete = Complete andalso Tail =:= [] andalso no_persistent_sessions(),
+                    emqx_mgmt_api:finalize_query(
+                        NResultAcc, emqx_mgmt_api:mark_complete(FQueryState, FComplete)
+                    );
+                {more, NResultAcc} when not Complete ->
+                    do_list_clients_cluster_query(Nodes, NQueryState, NResultAcc);
+                {more, NResultAcc} when Tail =/= [] ->
+                    do_list_clients_cluster_query(
+                        Tail, emqx_mgmt_api:reset_query_state(NQueryState), NResultAcc
+                    );
+                {more, NResultAcc} ->
+                    do_persistent_session_query(NResultAcc, NQueryState)
+            end
+    end.
+
+list_clients_node_query(Node, QString, Options) ->
+    case emqx_mgmt_api:parse_pager_params(QString) of
+        false ->
+            {error, page_limit_invalid};
+        Meta = #{} ->
+            {_CodCnt, NQString} = emqx_mgmt_api:parse_qstring(QString, ?CLIENT_QSCHEMA),
+            ResultAcc = emqx_mgmt_api:init_query_result(),
+            QueryState = emqx_mgmt_api:init_query_state(
+                ?CHAN_INFO_TAB, NQString, fun ?MODULE:qs2ms/2, Meta, Options
+            ),
+            Res = do_list_clients_node_query(Node, QueryState, ResultAcc),
+            emqx_mgmt_api:format_query_result(fun ?MODULE:format_channel_info/2, Meta, Res)
+    end.
+
+%% adapted from `emqx_mgmt_api:do_node_query'
+do_list_clients_node_query(
+    Node,
+    QueryState,
+    ResultAcc
+) ->
+    case emqx_mgmt_api:do_query(Node, QueryState) of
+        {error, Error} ->
+            {error, Node, Error};
+        {Rows, NQueryState = #{complete := Complete}} ->
+            case emqx_mgmt_api:accumulate_query_rows(Node, Rows, NQueryState, ResultAcc) of
+                {enough, NResultAcc} ->
+                    FComplete = Complete andalso no_persistent_sessions(),
+                    emqx_mgmt_api:finalize_query(
+                        NResultAcc, emqx_mgmt_api:mark_complete(NQueryState, FComplete)
+                    );
+                {more, NResultAcc} when Complete ->
+                    do_persistent_session_query(NResultAcc, NQueryState);
+                {more, NResultAcc} ->
+                    do_list_clients_node_query(Node, NQueryState, NResultAcc)
+            end
+    end.
+
+init_persistent_session_iterator() ->
+    emqx_persistent_session_ds_state:make_session_iterator().
+
+no_persistent_sessions() ->
+    case emqx_persistent_message:is_persistence_enabled() of
+        true ->
+            emqx_persistent_session_ds_state:is_end_of_table(init_persistent_session_iterator());
```

Comment:
Q: Why use separate function here while there's a simple comparison with `$end_of_table` down below?

Reply:
Good question.  In my first attempt, it was just like that.  But dialyzer complained about such a comparison here (because the iterator is opaque), so I had to introduce this function...

Curiously enough, it's not complaining about the use case in the function guard, in this same file.  :thinking:

Reply:
Maybe we could make that typespec:

```erlang
-opaque session_iterator_inner() :: emqx_persistent_session_ds:id().
-type session_iterator() ::  session_iterator_inner() | '$end_of_table'.
```

Reply:
This spec feels a bit odd to me: hard to make sense of a union of opaque type and non-opaque type. I think it's better either to use `is_end_of_table/1` everywhere, to turn it into a regular type, or to change iteration function's contract into something like `next(iterator()) -> {_Items, {more, iterator()} | fin}`.